storage_service/locations/models/package.py

```diff
@@ -522,10 +522,30 @@ def extract_file(self, relative_path='', extract_path=None):
             output_path = os.path.join(extract_path, basename)
 
         if self.is_compressed:
-            command = ['unar', '-force-overwrite', '-o', extract_path, full_path]
+            # The command used to extract the compressed file at
+            # full_path was, previously, universally::
+            #
+            #     $ unar -force-overwrite -o extract_path full_path
+            #
+            # The problem with this command is that unar treats __MACOSX .rsrc
+            # ("resource fork") files differently than 7z and tar do. 7z and
+            # tar convert these .rsrc files to ._-prefixed files. Similar
+            # behaviour with unar can be achieved by passing `-k hidden`.
+            # However, while a command like::
+            #
+            #     $ unar -force-overwrite -k hidden -o extract_path full_path
+            #
+            # preserves the .rsrc MACOSX files as ._-prefixed files, it does so
+            # differently than 7z/tar do: the resulting .-prefixed files have
+            # different sizes than those created via unar. This makes
+            # ``bag.validate`` choke.
+            if self.full_pointer_file_path:
+                compression = self._get_compression(self.full_pointer_file_path)
+            else:
+                compression = None  # no pointer file :. command will be unar
+            command = self._get_decompr_cmd(compression, extract_path, full_path)
             if relative_path:
                 command.append(relative_path)
```

Comment:
Yes, my investigation concluded  that 7z, unar and tar all allow for a final argument that specifies a relative path within the compressed directory. Please tell me if you think this is not the case.